Add users support for FreeBSD

xaxo05 opened the pull request:
Add users support for FreeBSD.

Use pw insread of useradd and usermod, user chpass instead of chpasswd, use rmuser instead of userdel. 

Implement fgetpwent() and fgetgrent() for FreeBSD.

Needs to be tested as I have only tested some cases myself for my personal needs.

cf-bottom commented:
Thanks for submitting a PR! Maybe @vpodzime can review this?

vpodzime commented:
Thanks for the contribution @xaxo05! Before I even get to a review, I can see that there are some issues in the commit history. Can you please de-duplicate the commits?

xaxo05 commented:
Sorry, I am not proficient git user, so bear with me, please.

 I did a "git reset --hard" on the branch and applied the patches obtained using "git show" by hand. After that I pushed the branch with the "--force" parameter. I hope the result is what you wanted. I did a "git rebase" a couple of times when I was splitting the code changes in separate commits, that might have broken the history before.

I can also send you the final state of the patched files and the patches themselves. I have those lying around as text files obtained from the "git show" command, so that I have the commit message as well.

vpodzime commented on cf-agent/Makefile.am:
```diff
@@ -124,7 +124,11 @@ if !BUILTIN_EXTENSIONS
   # Workaround for automake madness (try removing it if you want to know why).
-  cf_agent_CFLAGS = $(AM_CFLAGS)
+if FREEBSD
+    cf_agent_CFLAGS = $(AM_CFLAGS) -lutil
```
`-lutil` should definitely go to `CFLAGS`, it's a linker flag. So I guess this should actually go to `AM_LDFLAGS` above?

xaxo05 replied:
fixed

vpodzime commented on configure.ac:
```diff
@@ -929,15 +954,29 @@ AS_IF([test x$with_pam != xno],
+         [have_fgetpwent_fgetgrent=no]
+     )
+
+     AS_CASE(["$target_os"],
```
Please fix the indentation here to make it easier to see which macros are at the same level.

xaxo05 replied:
fixed

olehermanse commented on cf-agent/verify_users_pam.c:
```diff
@@ -918,6 +964,10 @@ static void TransformGidsToGroups(StringSet **list)
             StringSetAdd(new_list, xstrdup(data));
         }
+#ifdef __FreeBSD__
+        if (group_info)
```
Unnecessary `if`.

olehermanse commented on cf-agent/verify_users_pam.c:
```diff
@@ -995,6 +1045,11 @@ static bool VerifyIfUserNeedsModifs (const char *puser, const User *u, const str
         }
+
+#ifdef __FreeBSD__
+        if (group_info)
```
Unnecessary `if`.

olehermanse commented on cf-agent/verify_users_pam.c:
```diff
@@ -1403,6 +1458,42 @@ static bool DoModifyUser (const char *puser, const User *u, const struct passwd
 }
 
+#ifdef __FreeBSD__
+struct passwd *
```
Don't put the return type on a separate line.

olehermanse commented on cf-agent/verify_users_pam.c:
```diff
@@ -1403,6 +1458,42 @@ static bool DoModifyUser (const char *puser, const User *u, const struct passwd
+    pwd_scanflag = 0;
+
+    if (stream == NULL)
+            return NULL;
```
This function is very similar to the other one above. Please:

* Move the argument `if` check to the top.
* Combine variable declarations and initializations.
* Add curly braces to `if`s.

olehermanse commented on cf-agent/verify_users_pam.c:
```diff
@@ -1509,4 +1603,8 @@ void VerifyOneUsersPromise (const char *puser, const User *u, PromiseResult *res
         }
     }
+#ifdef __FreeBSD__
+    if (passwd_info)
```
Please remove this unnecessary `if`.

olehermanse commented on cf-agent/verify_users_pam.c:
```diff
@@ -1286,7 +1286,28 @@ static bool DoCreateUser(const char *puser, const User *u, enum cfopaction actio
+{
+    char cmd[CF_BUFSIZE];
+
+    strcpy (cmd, RMUSER);
```
Please remove the space after `strcpy`.

olehermanse commented on cf-agent/verify_users_pam.c:
```diff
@@ -1303,6 +1324,19 @@ static bool DoRemoveUser (const char *puser, enum cfopaction action)
+static bool DoRemoveUser (const char *puser, enum cfopaction action)
+{
+#if defined(HAVE_RMUSER)
+        return DoRemoveUserUsingRmuser(puser, action);
```
We use 4 spaces indentation level.

olehermanse commented on cf-agent/verify_users_pam.c:
```diff
@@ -1285,6 +1307,137 @@ static bool DoCreateUser(const char *puser, const User *u, enum cfopaction actio
+
+#ifdef HAVE_PW
+static bool DoCreateUserUsingPw(const char *puser, const User *u, enum cfopaction action,
+                         EvalContext *ctx, const Attributes *a, const Promise *pp)
```
There is a lot of duplication here. Can you factor out parts / most of this function to be shared by `DoCreateUserUsingUseradd` ?

olehermanse commented on cf-agent/verify_users_pam.c:
```diff
@@ -793,6 +796,55 @@ static bool SetAccountLockExpiration(const char *puser, bool lock)
+#endif
+
+#ifdef HAVE_PW
+static bool SetAccountLockExpirationUsingPw(const char *puser, bool lock)
```
Again try to avoid all the code duplication (copy-pasting) by having a shared helper function.

olehermanse commented on cf-agent/verify_users_pam.c:
```diff
@@ -864,9 +879,13 @@ static bool SetAccountLocked(const char *puser, const char *hash, bool lock)
+            if (IsHashLocked(hash))
             {
+#ifdef HAVE_PW
+                if (!ChangePassword(puser, &hash[8], PASSWORD_FORMAT_HASH))
```
Add a comment explaining where this `8` comes from / what it means.

olehermanse commented on cf-agent/verify_users_pam.c:
```diff
@@ -622,6 +630,46 @@ static bool ChangePasswordHashUsingLckpwdf(const char *puser, const char *passwo
+    Log(LOG_LEVEL_VERBOSE, "%s user '%s'. (command: '%s')", cap_action_msg, puser, cmd);
+
+    int status;
+    status = system(cmd);
```
Should be one line and const:

```
const int status = system(cmd); 
```

olehermanse commented on cf-agent/verify_users_pam.c:
```diff
@@ -668,6 +720,19 @@ static bool ChangePassword(const char *puser, const char *password, PasswordForm
+{
+#ifdef HAVE_PW
+    /* Accounts are locked by prepending "*LOCKED*" to the password
+     * hash on FreeBSD. */
```
Comment indicates that this is specific to FreeBSD, but macro is `HAVE_PW` which one is correct?

olehermanse replied:
> Regarding comment [1], the user management tool on FreeBSD is pw. The way pw locks user accounts is by prepending the password hash with LOCKED. This way the password hash does not match your hashed password, thus you are unable to login using your password. The string used to prepend the password hash can be arbitrary. If you stick to LOCKED, you will be able to unlock a user account with pw. The operating system, FreeBSD in particular, does not care about what string you use. This mechanism of altering the password hash of the user entry can also be applied to any other system. I can change the comment and replace FreeBSD with pw as there are a few other operating systems forked from FreeBSD. Some of the prominent ones are DragonflyBSD, FreeSBIE, FreeNAS, pfSense and m0n0wall.

> Would that change be OK?

@xaxo05 It's easier if you reply inside this thread, instead of adding more top level comments.

Regarding the change, something like this would be sufficient:

```
/* Accounts are locked by prepending "*LOCKED*" to the password
     * hash on FreeBSD and other systems using pw. */
```

It's just confusing when the macro and the comment seem to contradict each other. (Specific to pw or specific to freebsd).

olehermanse commented on cf-agent/verify_users_pam.c:
```diff
@@ -743,17 +810,71 @@ static bool SetAccountLockExpiration(const char *puser, bool lock)
+
+    char cmd[CF_BUFSIZE + strlen(puser)];
+
+    strcpy (cmd, PW);
```
No space after function name.

olehermanse commented on cf-agent/verify_users_pam.c:
```diff
@@ -743,17 +810,71 @@ static bool SetAccountLockExpiration(const char *puser, bool lock)
+        lock ? "Locking" : "Unlocking", puser, cmd);
+
+    int status;
+    status = system(cmd);
```
Should be one line and const:

```
const int status = system(cmd); 
```

vpodzime commented:
ping @xaxo05

xaxo05 commented:
I was on vacation, I will take a look at the latest comments and implement the changes.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cfengine/core?pullRequest=3589) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/cfengine/core?pullRequest=3589) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cfengine/core?pullRequest=3589) it.</sub>

xaxo05 commented:
# cat test.cf 
body common control
{
  bundlesequence => { "users_main" };
}

bundle agent users_main
{
  users:
    any::
      # remove existing account
      "news"
        policy => "absent";

      # create basic user
      "userA"
        policy => "present",
        description => "User userA",
        home_dir => "/home/userA",
        group_primary => "man",
        shell => "/bin/sh";

      # basic user with secondary groups
      "userB"
        policy => "present",
        description => "User userB",
        home_dir => "/home/userB",
        group_primary => "man",
        groups_secondary => {"video", "staff"},
        shell => "/bin/csh";

      # basic user with uid
      "userC"
        policy => "present",
        description => "User userC",
        home_dir => "/home/userC",
        group_primary => "man",
        shell => "/bin/csh",
        uid => "1357";

      # lock existing account
      "games"
        policy => "locked";

      # create locked account
      "userD"
        policy => "locked",
        description => "User userD",
        home_dir => "/home/userD",
        group_primary => "man",
        shell => "/bin/csh";

      # set plaintext password
      "userE"
        policy => "present",
        description => "User userE",
        home_dir => "/home/userE",
        group_primary => "man",
        shell => "/bin/csh",
        password => user_password_plain;

      # set hashed password
      "userF"
        policy => "present",
        description => "User userF",
        home_dir => "/home/userF",
        group_primary => "man",
        shell => "/bin/csh",
        password => user_password_hash;

      # create home directories
      "userG"
        policy => "present",
        description => "User userG",
        home_dir => "/home/userG",
        group_primary => "man",
        shell => "/bin/csh",
        home_bundle => setup_home_dir("userG");
}

bundle agent setup_home_dir(user)
{
  vars:
      "directories" slist =>
      {
        "/home/$(user)/.",
        "/home/$(user)/Maildir/.",
        "/home/$(user)/Maildir/cur/.",
        "/home/$(user)/Maildir/new/.",
        "/home/$(user)/Maildir/tmp/."
      };

  files:
      "/home/." create => "true";

      "$(directories)"
        create => "true",
        perms => perms_homedir("$(user)");
}

body perms perms_homedir(user)
{
        mode => "0700";
        owners => {"$(user)"};
}

body password user_password_plain
{
        format => "plaintext";
        data => "test";
}

body password user_password_hash
{
        format => "hash";
        data => "$6$XKc8afFsy9s6FZhY$CSgsDgjqxftMIRmFYokHNOwWkXAlhOGvdzVmURsd13R6nqxTpx.TNZrwVNqL3exPhewxE.LAbXBAf0EjTzF0A1";
}

xaxo05 commented:
# pw user show -a
root::0:0::0:0:Charlie &:/root:/bin/csh
toor:*:0:0::0:0:Bourne-again Superuser:/root:
daemon:*:1:1::0:0:Owner of many system processes:/root:/usr/sbin/nologin
operator:*:2:5::0:0:System &:/:/usr/sbin/nologin
bin:*:3:7::0:0:Binaries Commands and Source:/:/usr/sbin/nologin
tty:*:4:65533::0:0:Tty Sandbox:/:/usr/sbin/nologin
kmem:*:5:65533::0:0:KMem Sandbox:/:/usr/sbin/nologin
games:*:7:13::0:86400:Games pseudo-user:/:/usr/sbin/nologin
news:*:8:8::0:0:News Subsystem:/:/usr/sbin/nologin
man:*:9:9::0:0:Mister Man Pages:/usr/share/man:/usr/sbin/nologin
sshd:*:22:22::0:0:Secure Shell Daemon:/var/empty:/usr/sbin/nologin
smmsp:*:25:25::0:0:Sendmail Submission User:/var/spool/clientmqueue:/usr/sbin/nologin
mailnull:*:26:26::0:0:Sendmail Default User:/var/spool/mqueue:/usr/sbin/nologin
bind:*:53:53::0:0:Bind Sandbox:/:/usr/sbin/nologin
unbound:*:59:59::0:0:Unbound DNS Resolver:/var/unbound:/usr/sbin/nologin
proxy:*:62:62::0:0:Packet Filter pseudo-user:/nonexistent:/usr/sbin/nologin
_pflogd:*:64:64::0:0:pflogd privsep user:/var/empty:/usr/sbin/nologin
_dhcp:*:65:65::0:0:dhcp programs:/var/empty:/usr/sbin/nologin
uucp:*:66:66::0:0:UUCP pseudo-user:/var/spool/uucppublic:/usr/local/libexec/uucp/uucico
pop:*:68:6::0:0:Post Office Owner:/nonexistent:/usr/sbin/nologin
auditdistd:*:78:77::0:0:Auditdistd unprivileged user:/var/empty:/usr/sbin/nologin
www:*:80:80::0:0:World Wide Web Owner:/nonexistent:/usr/sbin/nologin
ntpd:*:123:123::0:0:NTP Daemon:/var/db/ntp:/usr/sbin/nologin
_ypldap:*:160:160::0:0:YP LDAP unprivileged user:/var/empty:/usr/sbin/nologin
hast:*:845:845::0:0:HAST unprivileged user:/var/empty:/usr/sbin/nologin
nobody:*:65534:65534::0:0:Unprivileged user:/nonexistent:/usr/sbin/nologin
user:$6$5NjmI3secLRcvGjd$P.xjJTZIMEBq6gJjHNEDqndCS0OR5bhoHgfWzn9e4Wj5O4Lc.SwQ6XXi2ehSIctRak8EPKec94OPc1IY9JZyc1:1001:1001::0:0:User &:/home/user:/bin/sh
git_daemon:*:964:964::0:0:git daemon:/nonexistent:/usr/sbin/nologin
_tss:*:601:601:daemon:0:0:TrouSerS user:/var/empty:/usr/sbin/nologin
# /usr/local/bin/cf-agent -IK -f ./test.cf -v
 verbose: Could not open extension plugin 'cfengine-enterprise.so' from '/var/cfengine/lib/cfengine-enterprise.so': (not installed)
 verbose:  CFEngine Core 3.13.0
 verbose: ----------------------------------------------------------------
 verbose:  Initialization preamble 
 verbose: ----------------------------------------------------------------
 verbose: Default port for cfengine is 5308
 verbose: Work directory is /var/cfengine
 verbose: Making sure that internal directories are private...
 verbose: Checking integrity of the trusted workdir
 verbose: Checking integrity of the state database
 verbose: Checking integrity of the module directory
 verbose: Checking integrity of the PKI directory
 verbose: Couldn't find a private key at '/var/cfengine/ppkeys/localhost.priv', use cf-key to get one. (fopen: No such file or directory)
 verbose: Could not open file '/var/cfengine/policy_server.dat' (fopen: No such file or directory)
 verbose: Reference time set to 'Wed Oct  9 00:18:57 2019'
 verbose: CFEngine Core 3.13.0 - ready
 verbose: ----------------------------------------------------------------
 verbose:  Environment discovery 
 verbose: ----------------------------------------------------------------
 verbose: Host name is: freebsd
 verbose: Operating System Type is freebsd
 verbose: Operating System Release is 12.0-RELEASE
 verbose: Architecture = amd64
 verbose: CFEngine detected operating system description is freebsd
 verbose: The time is now Wed Oct  9 00:18:57 2019
 verbose: Additional hard class defined as: 64_bit
 verbose: Additional hard class defined as: freebsd_12_0_RELEASE
 verbose: Additional hard class defined as: freebsd_amd64
 verbose: Additional hard class defined as: freebsd_amd64_12_0_RELEASE
 verbose: GNU autoconf class from compile time: compiled_on_freebsd12_0
 verbose: Hostname lookup failed on node name 'freebsd'
 verbose: No interface exception file /var/cfengine/inputs/ignore_interfaces.rx
 verbose: Interface 1: em0
 verbose: Interface 2: em0
 verbose: Interface 3: em0
 verbose: IP address of host set to 10.0.2.15
 verbose: Interface 4: em1
 verbose: Interface 5: em1
 verbose: Interface 6: em1
 verbose: Interface 7: lo0
 verbose: Interface 8: lo0
 verbose: Interface 9: lo0
 verbose: Interface 10: lo0
 verbose: Trying to locate my IPv6 address
 verbose: Looking for environment from cf-monitord...
 verbose: Unable to detect environment from cf-monitord
 verbose: Found 1 processor
 verbose: Loading persistent classes
 verbose: Could not open file '/var/cfengine/policy_server.dat' (fopen: No such file or directory)
 verbose: This agent is not bootstrapped - can't find policy_server.dat in: /var/cfengine
 verbose: Loading JSON augments from './def.json' (input dir '.', input file './test.cf'
 verbose: could not load JSON augments from './def.json'
 verbose: Input file is outside default repository, validating it
 verbose: Input file './test.cf' has changed since the last policy read attempt (file is newer than previous)
 verbose: Input file is changed since last validation, validating it
 verbose: Verifying the syntax of the inputs...
 verbose: Checking policy with command '"/usr/local/bin/cf-promises" -c "./test.cf"'
 verbose: Saved policy validated marker file '/var/cfengine/state/cf_promises_validated'
 verbose: ----------------------------------------------------------------
 verbose:  Loading policy 
 verbose: ----------------------------------------------------------------
 verbose: BEGIN parsing file: ./test.cf
 verbose: END   parsing file: ./test.cf
 verbose: Running full policy integrity checks
 verbose: ----------------------------------------------------------------
 verbose: PREFIX LEGEND:
 verbose:  V: variable or parameter new definition in scope
 verbose:  C: class/context new definition 
 verbose:  B: bundle start/end execution marker
 verbose:  P: promise execution output 
 verbose:  A: accounting output 
 verbose:  T: time measurement for stated object (promise or bundle)
 verbose: ----------------------------------------------------------------
 verbose: ----------------------------------------------------------------
 verbose: BEGIN Discovered hard classes:
 verbose: C: discovered hard class 10_0_2_15
 verbose: C: discovered hard class 127_0_0_1
 verbose: C: discovered hard class 192_168_56_105
 verbose: C: discovered hard class 1_cpu
 verbose: C: discovered hard class 64_bit
 verbose: C: discovered hard class Day9
 verbose: C: discovered hard class GMT_Day9
 verbose: C: discovered hard class GMT_Hr0
 verbose: C: discovered hard class GMT_Hr00
 verbose: C: discovered hard class GMT_Hr00_Q2
 verbose: C: discovered hard class GMT_Lcycle_0
 verbose: C: discovered hard class GMT_Min15_20
 verbose: C: discovered hard class GMT_Min18
 verbose: C: discovered hard class GMT_Night
 verbose: C: discovered hard class GMT_October
 verbose: C: discovered hard class GMT_Q2
 verbose: C: discovered hard class GMT_Wednesday
 verbose: C: discovered hard class GMT_Yr2019
 verbose: C: discovered hard class Hr0
 verbose: C: discovered hard class Hr00
 verbose: C: discovered hard class Hr00_Q2
 verbose: C: discovered hard class Lcycle_0
 verbose: C: discovered hard class Min15_20
 verbose: C: discovered hard class Min18
 verbose: C: discovered hard class Night
 verbose: C: discovered hard class October
 verbose: C: discovered hard class Q2
 verbose: C: discovered hard class Wednesday
 verbose: C: discovered hard class Yr2019
 verbose: C: discovered hard class agent
 verbose: C: discovered hard class amd64
 verbose: C: discovered hard class any
 verbose: C: discovered hard class cfengine
 verbose: C: discovered hard class cfengine_3
 verbose: C: discovered hard class cfengine_3_13
 verbose: C: discovered hard class cfengine_3_13_0
 verbose: C: discovered hard class community_edition
 verbose: C: discovered hard class compiled_on_freebsd12_0
 verbose: C: discovered hard class feature
 verbose: C: discovered hard class feature_curl
 verbose: C: discovered hard class feature_def
 verbose: C: discovered hard class feature_def_json
 verbose: C: discovered hard class feature_def_json_preparse
 verbose: C: discovered hard class feature_tls
 verbose: C: discovered hard class feature_tls_1
 verbose: C: discovered hard class feature_tls_1_0
 verbose: C: discovered hard class feature_tls_1_1
 verbose: C: discovered hard class feature_tls_1_2
 verbose: C: discovered hard class feature_tls_1_3
 verbose: C: discovered hard class feature_xml
 verbose: C: discovered hard class freebsd
 verbose: C: discovered hard class freebsd_12
 verbose: C: discovered hard class freebsd_12_0
 verbose: C: discovered hard class freebsd_12_0_RELEASE
 verbose: C: discovered hard class freebsd_amd64
 verbose: C: discovered hard class freebsd_amd64_12_0_RELEASE
 verbose: C: discovered hard class freebsd_amd64_12_0_RELEASE_FreeBSD_12_0_RELEASE_r341666_GENERIC
 verbose: C: discovered hard class inform_mode
 verbose: C: discovered hard class ipv4_10
 verbose: C: discovered hard class ipv4_10_0
 verbose: C: discovered hard class ipv4_10_0_2
 verbose: C: discovered hard class ipv4_10_0_2_15
 verbose: C: discovered hard class ipv4_127
 verbose: C: discovered hard class ipv4_127_0
 verbose: C: discovered hard class ipv4_127_0_0
 verbose: C: discovered hard class ipv4_127_0_0_1
 verbose: C: discovered hard class ipv4_192
 verbose: C: discovered hard class ipv4_192_168
 verbose: C: discovered hard class ipv4_192_168_56
 verbose: C: discovered hard class ipv4_192_168_56_105
 verbose: C: discovered hard class mac_08_00_27_8e_ae_90
 verbose: C: discovered hard class mac_08_00_27_c9_63_9d
 verbose: C: discovered hard class net_iface_em0
 verbose: C: discovered hard class net_iface_em1
 verbose: C: discovered hard class net_iface_lo0
 verbose: C: discovered hard class verbose_mode
 verbose: END Discovered hard classes
 verbose: ----------------------------------------------------------------
 verbose: ----------------------------------------------------------------
 verbose:  Preliminary variable/class-context convergence 
 verbose: ----------------------------------------------------------------
 verbose: No public/private key pair is loaded, please create one using cf-key
 verbose: ----------------------------------------------------------------
 verbose:  Begin policy/promise evaluation 
 verbose: ----------------------------------------------------------------
 verbose: Using bundlesequence =>  {"users_main"}
 verbose: B: *****************************************************************
 verbose: B: BEGIN bundle users_main
 verbose: B: *****************************************************************
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_11' of type "users" (pass 1)
 verbose: P:    Promiser/affected object: 'news'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'news'[1]
 verbose: Removing user 'news'. (command: '/usr/sbin/rmuser -y news')
Removing user (news): home passwd.
    info: User promise repaired
 verbose: A: Promise REPAIRED
 verbose: P: END users promise (news)
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_15' of type "users" (pass 1)
 verbose: P:    Promiser/affected object: 'userA'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userA'[1]
 verbose: Creating user 'userA'. (command: '/usr/sbin/pw useradd userA -c "User userA" -g "man" -d "/home/userA" -s "/bin/sh"')
 verbose: Changing password hash for user 'userA'. (command: '/usr/bin/chpass -p x userA')
 verbose: Changing user 'userA'. (command: '/usr/bin/chpass -p x userA')
chpass: user information updated
    info: User promise repaired
 verbose: A: Promise REPAIRED
 verbose: P: END users promise (userA)
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_23' of type "users" (pass 1)
 verbose: P:    Promiser/affected object: 'userB'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userB'[1]
 verbose: Creating user 'userB'. (command: '/usr/sbin/pw useradd userB -c "User userB" -g "man" -G "video,staff" -d "/home/userB" -s "/bin/csh"')
 verbose: Changing password hash for user 'userB'. (command: '/usr/bin/chpass -p x userB')
 verbose: Changing user 'userB'. (command: '/usr/bin/chpass -p x userB')
chpass: user information updated
    info: User promise repaired
 verbose: A: Promise REPAIRED
 verbose: P: END users promise (userB)
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_32' of type "users" (pass 1)
 verbose: P:    Promiser/affected object: 'userC'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userC'[1]
 verbose: Creating user 'userC'. (command: '/usr/sbin/pw useradd userC -u "1357" -c "User userC" -g "man" -d "/home/userC" -s "/bin/csh"')
 verbose: Changing password hash for user 'userC'. (command: '/usr/bin/chpass -p x userC')
 verbose: Changing user 'userC'. (command: '/usr/bin/chpass -p x userC')
chpass: user information updated
    info: User promise repaired
 verbose: A: Promise REPAIRED
 verbose: P: END users promise (userC)
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_41' of type "users" (pass 1)
 verbose: P:    Promiser/affected object: 'games'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'games'[1]
 verbose: Getting user 'games' password hash from passwd database.
 verbose: Getting user 'games' password hash from passwd database.
 verbose: Changing password hash for user 'games'. (command: '/usr/bin/chpass -p *LOCKED** games')
 verbose: Changing user 'games'. (command: '/usr/bin/chpass -p *LOCKED** games')
chpass: user information updated
 verbose: Locking user 'games' by setting expiry date. (command: '/usr/sbin/pw usermod games -e "02-Jan-1970" ')
    info: User promise repaired
 verbose: A: Promise REPAIRED
 verbose: P: END users promise (games)
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_45' of type "users" (pass 1)
 verbose: P:    Promiser/affected object: 'userD'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userD'[1]
 verbose: Creating user 'userD'. (command: '/usr/sbin/pw useradd userD -c "User userD" -g "man" -d "/home/userD" -s "/bin/csh"')
 verbose: Changing password hash for user 'userD'. (command: '/usr/bin/chpass -p x userD')
 verbose: Changing user 'userD'. (command: '/usr/bin/chpass -p x userD')
chpass: user information updated
 verbose: Changing password hash for user 'userD'. (command: '/usr/bin/chpass -p *LOCKED*x userD')
 verbose: Changing user 'userD'. (command: '/usr/bin/chpass -p *LOCKED*x userD')
chpass: user information updated
 verbose: Locking user 'userD' by setting expiry date. (command: '/usr/sbin/pw usermod userD -e "02-Jan-1970" ')
    info: User promise repaired
 verbose: A: Promise REPAIRED
 verbose: P: END users promise (userD)
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_53' of type "users" (pass 1)
 verbose: P:    Promiser/affected object: 'userE'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userE'[1]
 verbose: Creating user 'userE'. (command: '/usr/sbin/pw useradd userE -c "User userE" -g "man" -d "/home/userE" -s "/bin/csh"')
 verbose: Changing password hash for user 'userE'. (command: '/usr/bin/chpass -p x userE')
 verbose: Changing user 'userE'. (command: '/usr/bin/chpass -p x userE')
chpass: user information updated
 verbose: Changing password for user 'userE'.
    info: User promise repaired
 verbose: A: Promise REPAIRED
 verbose: P: END users promise (userE)
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_62' of type "users" (pass 1)
 verbose: P:    Promiser/affected object: 'userF'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userF'[1]
 verbose: Creating user 'userF'. (command: '/usr/sbin/pw useradd userF -c "User userF" -g "man" -d "/home/userF" -s "/bin/csh"')
 verbose: Changing password hash for user 'userF'. (command: '/usr/bin/chpass -p x userF')
 verbose: Changing user 'userF'. (command: '/usr/bin/chpass -p x userF')
chpass: user information updated
 verbose: Changing password hash for user 'userF'. (command: '/usr/bin/chpass -p $6$XKc8afFsy9s6FZhY$CSgsDgjqxftMIRmFYokHNOwWkXAlhOGvdzVmURsd13R6nqxTpx.TNZrwVNqL3exPhewxE.LAbXBAf0EjTzF0A1 userF')
 verbose: Changing user 'userF'. (command: '/usr/bin/chpass -p $6$XKc8afFsy9s6FZhY$CSgsDgjqxftMIRmFYokHNOwWkXAlhOGvdzVmURsd13R6nqxTpx.TNZrwVNqL3exPhewxE.LAbXBAf0EjTzF0A1 userF')
chpass: user information updated
    info: User promise repaired
 verbose: A: Promise REPAIRED
 verbose: P: END users promise (userF)
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_71' of type "users" (pass 1)
 verbose: P:    Promiser/affected object: 'userG'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userG'[1]
 verbose: Creating user 'userG'. (command: '/usr/sbin/pw useradd userG -c "User userG" -g "man" -d "/home/userG" -s "/bin/csh"')
 verbose: Changing password hash for user 'userG'. (command: '/usr/bin/chpass -p x userG')
 verbose: Changing user 'userG'. (command: '/usr/bin/chpass -p x userG')
chpass: user information updated
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_71' of type "users" (pass 1)
 verbose: P:    Promiser/affected object: 'userG'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userG'[1]
 verbose: B: *****************************************************************
 verbose: B: BEGIN bundle setup_home_dir( {"userG"})
 verbose: B: *****************************************************************
 verbose: V:     +  Private parameter: 'user' in scope 'setup_home_dir' (type: s) in pass 1
 verbose: V: .........................................................
 verbose: V: BEGIN variables (pass 1)
 verbose: V:     Computing value of 'directories'
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_93' of type "files" (pass 1)
 verbose: P:    Promiser/affected object: '/home/.'
 verbose: P:    From parameterized bundle: setup_home_dir( {"userG"})
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userG'/default/setup_home_dir/files/'/home/.'[1]
 verbose: Using literal pathtype for '/home/.'
 verbose: Additional promise info: source path './test.cf' at line 93
 verbose: File '/home/.' exists as promised
 verbose: A: Promise was KEPT
 verbose: P: END files promise (/home)
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_95' of type "files" (pass 1)
 verbose: P:    Promiser/affected object: '/home/userG/.'
 verbose: P:    From parameterized bundle: setup_home_dir( {"userG"})
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userG'/default/setup_home_dir/files/'/home/userG/.'[1]
 verbose: Using literal pathtype for '/home/userG/.'
    info: Created directory '/home/userG/.'
 verbose: Handling file existence constraints on '/home/userG'
    info: Owner of '/home/userG' was 0, setting to 1361
    info: Object '/home/userG' had permission 0755, changed it to 0700
 verbose: A: Promise REPAIRED
 verbose: P: END files promise (/home/userG)
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_95' of type "files" (pass 1)
 verbose: P:    Promiser/affected object: '/home/userG/Maildir/.'
 verbose: P:    From parameterized bundle: setup_home_dir( {"userG"})
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userG'/default/setup_home_dir/files/'/home/userG/Maildir/.'[2]
 verbose: Using literal pathtype for '/home/userG/Maildir/.'
    info: Created directory '/home/userG/Maildir/.'
 verbose: Handling file existence constraints on '/home/userG/Maildir'
    info: Owner of '/home/userG/Maildir' was 0, setting to 1361
    info: Object '/home/userG/Maildir' had permission 0755, changed it to 0700
 verbose: A: Promise REPAIRED
 verbose: P: END files promise (/home/userG/Maildir)
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_95' of type "files" (pass 1)
 verbose: P:    Promiser/affected object: '/home/userG/Maildir/cur/.'
 verbose: P:    From parameterized bundle: setup_home_dir( {"userG"})
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userG'/default/setup_home_dir/files/'/home/userG/Maildir/cur/.'[3]
 verbose: Using literal pathtype for '/home/userG/Maildir/cur/.'
    info: Created directory '/home/userG/Maildir/cur/.'
 verbose: Handling file existence constraints on '/home/userG/Maildir/cur'
    info: Owner of '/home/userG/Maildir/cur' was 0, setting to 1361
    info: Object '/home/userG/Maildir/cur' had permission 0755, changed it to 0700
 verbose: A: Promise REPAIRED
 verbose: P: END files promise (/home/userG/Maildir/cur)
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_95' of type "files" (pass 1)
 verbose: P:    Promiser/affected object: '/home/userG/Maildir/new/.'
 verbose: P:    From parameterized bundle: setup_home_dir( {"userG"})
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userG'/default/setup_home_dir/files/'/home/userG/Maildir/new/.'[4]
 verbose: Using literal pathtype for '/home/userG/Maildir/new/.'
    info: Created directory '/home/userG/Maildir/new/.'
 verbose: Handling file existence constraints on '/home/userG/Maildir/new'
    info: Owner of '/home/userG/Maildir/new' was 0, setting to 1361
    info: Object '/home/userG/Maildir/new' had permission 0755, changed it to 0700
 verbose: A: Promise REPAIRED
 verbose: P: END files promise (/home/userG/Maildir/new)
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_95' of type "files" (pass 1)
 verbose: P:    Promiser/affected object: '/home/userG/Maildir/tmp/.'
 verbose: P:    From parameterized bundle: setup_home_dir( {"userG"})
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userG'/default/setup_home_dir/files/'/home/userG/Maildir/tmp/.'[5]
 verbose: Using literal pathtype for '/home/userG/Maildir/tmp/.'
    info: Created directory '/home/userG/Maildir/tmp/.'
 verbose: Handling file existence constraints on '/home/userG/Maildir/tmp'
    info: Owner of '/home/userG/Maildir/tmp' was 0, setting to 1361
    info: Object '/home/userG/Maildir/tmp' had permission 0755, changed it to 0700
 verbose: A: Promise REPAIRED
 verbose: P: END files promise (/home/userG/Maildir/tmp)
 verbose: V: .........................................................
 verbose: V: BEGIN variables (pass 2)
 verbose: V:     Computing value of 'directories'
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_93' of type "files" (pass 2)
 verbose: P:    Promiser/affected object: '/home/.'
 verbose: P:    From parameterized bundle: setup_home_dir( {"userG"})
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userG'/default/setup_home_dir/files/'/home/.'[1]
 verbose: Using literal pathtype for '/home/.'
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_95' of type "files" (pass 2)
 verbose: P:    Promiser/affected object: '/home/userG/.'
 verbose: P:    From parameterized bundle: setup_home_dir( {"userG"})
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userG'/default/setup_home_dir/files/'/home/userG/.'[1]
 verbose: Using literal pathtype for '/home/userG/.'
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_95' of type "files" (pass 2)
 verbose: P:    Promiser/affected object: '/home/userG/Maildir/.'
 verbose: P:    From parameterized bundle: setup_home_dir( {"userG"})
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userG'/default/setup_home_dir/files/'/home/userG/Maildir/.'[2]
 verbose: Using literal pathtype for '/home/userG/Maildir/.'
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_95' of type "files" (pass 2)
 verbose: P:    Promiser/affected object: '/home/userG/Maildir/cur/.'
 verbose: P:    From parameterized bundle: setup_home_dir( {"userG"})
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userG'/default/setup_home_dir/files/'/home/userG/Maildir/cur/.'[3]
 verbose: Using literal pathtype for '/home/userG/Maildir/cur/.'
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_95' of type "files" (pass 2)
 verbose: P:    Promiser/affected object: '/home/userG/Maildir/new/.'
 verbose: P:    From parameterized bundle: setup_home_dir( {"userG"})
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userG'/default/setup_home_dir/files/'/home/userG/Maildir/new/.'[4]
 verbose: Using literal pathtype for '/home/userG/Maildir/new/.'
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_95' of type "files" (pass 2)
 verbose: P:    Promiser/affected object: '/home/userG/Maildir/tmp/.'
 verbose: P:    From parameterized bundle: setup_home_dir( {"userG"})
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userG'/default/setup_home_dir/files/'/home/userG/Maildir/tmp/.'[5]
 verbose: Using literal pathtype for '/home/userG/Maildir/tmp/.'
 verbose: V: .........................................................
 verbose: V: BEGIN variables (pass 3)
 verbose: V:     Computing value of 'directories'
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_93' of type "files" (pass 3)
 verbose: P:    Promiser/affected object: '/home/.'
 verbose: P:    From parameterized bundle: setup_home_dir( {"userG"})
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userG'/default/setup_home_dir/files/'/home/.'[1]
 verbose: Using literal pathtype for '/home/.'
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_95' of type "files" (pass 3)
 verbose: P:    Promiser/affected object: '/home/userG/.'
 verbose: P:    From parameterized bundle: setup_home_dir( {"userG"})
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userG'/default/setup_home_dir/files/'/home/userG/.'[1]
 verbose: Using literal pathtype for '/home/userG/.'
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_95' of type "files" (pass 3)
 verbose: P:    Promiser/affected object: '/home/userG/Maildir/.'
 verbose: P:    From parameterized bundle: setup_home_dir( {"userG"})
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userG'/default/setup_home_dir/files/'/home/userG/Maildir/.'[2]
 verbose: Using literal pathtype for '/home/userG/Maildir/.'
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_95' of type "files" (pass 3)
 verbose: P:    Promiser/affected object: '/home/userG/Maildir/cur/.'
 verbose: P:    From parameterized bundle: setup_home_dir( {"userG"})
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userG'/default/setup_home_dir/files/'/home/userG/Maildir/cur/.'[3]
 verbose: Using literal pathtype for '/home/userG/Maildir/cur/.'
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_95' of type "files" (pass 3)
 verbose: P:    Promiser/affected object: '/home/userG/Maildir/new/.'
 verbose: P:    From parameterized bundle: setup_home_dir( {"userG"})
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userG'/default/setup_home_dir/files/'/home/userG/Maildir/new/.'[4]
 verbose: Using literal pathtype for '/home/userG/Maildir/new/.'
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_95' of type "files" (pass 3)
 verbose: P:    Promiser/affected object: '/home/userG/Maildir/tmp/.'
 verbose: P:    From parameterized bundle: setup_home_dir( {"userG"})
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userG'/default/setup_home_dir/files/'/home/userG/Maildir/tmp/.'[5]
 verbose: Using literal pathtype for '/home/userG/Maildir/tmp/.'
 verbose: A: ...................................................
 verbose: A: Bundle Accounting Summary for 'setup_home_dir' in namespace default
 verbose: A: Promises kept in 'setup_home_dir' = 1
 verbose: A: Promises not kept in 'setup_home_dir' = 0
 verbose: A: Promises repaired in 'setup_home_dir' = 15
 verbose: A: Aggregate compliance (promises kept/repaired) for bundle 'setup_home_dir' = 100.0%
 verbose: A: ...................................................
 verbose: Additional promise info: source path './test.cf' at line 71
 verbose: Method 'setup_home_dir' invoked repairs
 verbose: B: *****************************************************************
 verbose: B: END bundle setup_home_dir
 verbose: B: *****************************************************************
    info: User promise repaired
 verbose: A: Promise REPAIRED
 verbose: P: END users promise (userG)
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_11' of type "users" (pass 2)
 verbose: P:    Promiser/affected object: 'news'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'news'[1]
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_15' of type "users" (pass 2)
 verbose: P:    Promiser/affected object: 'userA'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userA'[1]
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_23' of type "users" (pass 2)
 verbose: P:    Promiser/affected object: 'userB'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userB'[1]
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_32' of type "users" (pass 2)
 verbose: P:    Promiser/affected object: 'userC'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userC'[1]
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_41' of type "users" (pass 2)
 verbose: P:    Promiser/affected object: 'games'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'games'[1]
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_45' of type "users" (pass 2)
 verbose: P:    Promiser/affected object: 'userD'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userD'[1]
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_53' of type "users" (pass 2)
 verbose: P:    Promiser/affected object: 'userE'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userE'[1]
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_62' of type "users" (pass 2)
 verbose: P:    Promiser/affected object: 'userF'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userF'[1]
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_71' of type "users" (pass 2)
 verbose: P:    Promiser/affected object: 'userG'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userG'[1]
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_11' of type "users" (pass 3)
 verbose: P:    Promiser/affected object: 'news'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'news'[1]
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_15' of type "users" (pass 3)
 verbose: P:    Promiser/affected object: 'userA'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userA'[1]
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_23' of type "users" (pass 3)
 verbose: P:    Promiser/affected object: 'userB'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userB'[1]
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_32' of type "users" (pass 3)
 verbose: P:    Promiser/affected object: 'userC'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userC'[1]
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_41' of type "users" (pass 3)
 verbose: P:    Promiser/affected object: 'games'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'games'[1]
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_45' of type "users" (pass 3)
 verbose: P:    Promiser/affected object: 'userD'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userD'[1]
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_53' of type "users" (pass 3)
 verbose: P:    Promiser/affected object: 'userE'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userE'[1]
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_62' of type "users" (pass 3)
 verbose: P:    Promiser/affected object: 'userF'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userF'[1]
 verbose: P: .........................................................
 verbose: P: BEGIN promise 'promise_test_cf_71' of type "users" (pass 3)
 verbose: P:    Promiser/affected object: 'userG'
 verbose: P:    Part of bundle: users_main
 verbose: P:    Base context class: any
 verbose: P:    Stack path: /default/users_main/users/'userG'[1]
 verbose: A: ...................................................
 verbose: A: Bundle Accounting Summary for 'users_main' in namespace default
 verbose: A: Promises kept in 'users_main' = 1
 verbose: A: Promises not kept in 'users_main' = 0
 verbose: A: Promises repaired in 'users_main' = 25
 verbose: A: Aggregate compliance (promises kept/repaired) for bundle 'users_main' = 100.0%
 verbose: A: ...................................................
 verbose: B: *****************************************************************
 verbose: B: END bundle users_main
 verbose: B: *****************************************************************
 verbose: No lock purging scheduled
 verbose: Logging total compliance, total 'Outcome of version (not specified) (agent-0): Promises observed to be kept 3.85%, Promises repaired 96.15%, Promises not repaired 0.00%'
# pw user show -a
root::0:0::0:0:Charlie &:/root:/bin/csh
toor:*:0:0::0:0:Bourne-again Superuser:/root:
daemon:*:1:1::0:0:Owner of many system processes:/root:/usr/sbin/nologin
operator:*:2:5::0:0:System &:/:/usr/sbin/nologin
bin:*:3:7::0:0:Binaries Commands and Source:/:/usr/sbin/nologin
tty:*:4:65533::0:0:Tty Sandbox:/:/usr/sbin/nologin
kmem:*:5:65533::0:0:KMem Sandbox:/:/usr/sbin/nologin
games:*LOCKED**:7:13::0:86400:Games pseudo-user:/:/usr/sbin/nologin
man:*:9:9::0:0:Mister Man Pages:/usr/share/man:/usr/sbin/nologin
sshd:*:22:22::0:0:Secure Shell Daemon:/var/empty:/usr/sbin/nologin
smmsp:*:25:25::0:0:Sendmail Submission User:/var/spool/clientmqueue:/usr/sbin/nologin
mailnull:*:26:26::0:0:Sendmail Default User:/var/spool/mqueue:/usr/sbin/nologin
bind:*:53:53::0:0:Bind Sandbox:/:/usr/sbin/nologin
unbound:*:59:59::0:0:Unbound DNS Resolver:/var/unbound:/usr/sbin/nologin
proxy:*:62:62::0:0:Packet Filter pseudo-user:/nonexistent:/usr/sbin/nologin
_pflogd:*:64:64::0:0:pflogd privsep user:/var/empty:/usr/sbin/nologin
_dhcp:*:65:65::0:0:dhcp programs:/var/empty:/usr/sbin/nologin
uucp:*:66:66::0:0:UUCP pseudo-user:/var/spool/uucppublic:/usr/local/libexec/uucp/uucico
pop:*:68:6::0:0:Post Office Owner:/nonexistent:/usr/sbin/nologin
auditdistd:*:78:77::0:0:Auditdistd unprivileged user:/var/empty:/usr/sbin/nologin
www:*:80:80::0:0:World Wide Web Owner:/nonexistent:/usr/sbin/nologin
ntpd:*:123:123::0:0:NTP Daemon:/var/db/ntp:/usr/sbin/nologin
_ypldap:*:160:160::0:0:YP LDAP unprivileged user:/var/empty:/usr/sbin/nologin
hast:*:845:845::0:0:HAST unprivileged user:/var/empty:/usr/sbin/nologin
nobody:*:65534:65534::0:0:Unprivileged user:/nonexistent:/usr/sbin/nologin
user:$6$5NjmI3secLRcvGjd$P.xjJTZIMEBq6gJjHNEDqndCS0OR5bhoHgfWzn9e4Wj5O4Lc.SwQ6XXi2ehSIctRak8EPKec94OPc1IY9JZyc1:1001:1001::0:0:User &:/home/user:/bin/sh
git_daemon:*:964:964::0:0:git daemon:/nonexistent:/usr/sbin/nologin
_tss:*:601:601:daemon:0:0:TrouSerS user:/var/empty:/usr/sbin/nologin
userA:x:1002:9::0:0:User userA:/home/userA:/bin/sh
userB:x:1003:9::0:0:User userB:/home/userB:/bin/csh
userC:x:1357:9::0:0:User userC:/home/userC:/bin/csh
userD:*LOCKED*x:1358:9::0:86400:User userD:/home/userD:/bin/csh
userE:$6$tbeureaV4XhcHwHV$qg1je224N66q.KZk5QZcwRo13WPan6U.UCCdlAg4QTNcWu7LBIIAoXk91h3SjpKgNabtE3lKUu.vUvJpwuTin0:1359:9::0:0:User userE:/home/userE:/bin/csh
userF:.TNZrwVNqL3exPhewxE.LAbXBAf0EjTzF0A1:1360:9::0:0:User userF:/home/userF:/bin/csh
userG:x:1361:9::0:0:User userG:/home/userG:/bin/csh
# ls /home/
user    userG
# ls -la /home/ 
total 16
drwxr-xr-x   4 root   wheel  512 Oct  9 00:18 .
drwxr-xr-x  15 root   wheel  512 Oct  8 23:39 ..
drwxr-xr-x   3 user   user   512 Oct  8 23:49 user
drwx------   3 userG  wheel  512 Oct  9 00:18 userG

xaxo05 commented:
The hash for userF is not set properly and does not work, I will have to look into that, otherwise the rest looks good to me on first sight. A "-y" should be added to the rmuser command to skip asking for confirmation. I will commit that soon. I will also look at your other comments.

xaxo05 commented:
It just came to my mind, after the test I have created, that I have to remove the user using "pw userdel". This way I can remove only the user entry, as rmuser -y removes the user's home directory as well. According to [1] the user's home directory should not be touched. So that has to be changed.

1: https://docs.cfengine.com/docs/3.12/reference-promise-types-users.html

xaxo05 commented:
Regarding comment [1], the user management tool on FreeBSD is pw. The way pw locks user accounts is by prepending the password hash with *LOCKED*. This way the password hash does not match your hashed password, thus you are unable to login using your password. The string used to prepend the password hash can be arbitrary. If you stick to *LOCKED*, you will be able to unlock a user account with pw. The operating system, FreeBSD in particular, does not care about what string you use. This mechanism of altering the password hash of the user entry can also be applied to any other system. I can change the comment and replace FreeBSD with pw as there are a few other operating systems forked from FreeBSD. Some of the prominent ones are DragonflyBSD, FreeSBIE, FreeNAS, pfSense and m0n0wall.

Would that change be OK?

1: https://github.com/cfengine/core/pull/3589#discussion_r307242066

xaxo05 commented:
I have just fixed the password set/change by supplying a hash in commit c7372d18aa293b4e60a0997a1fb3f8a4e751e6d2. I see that there is no shell escape of any of the arguments supplied to the system() call. I needed to use single quotes to avoid the shell interpreting the dollar signs in the password hash. Maybe it would be a good idea to think about escaping all of the user supplied arguments somewhere? Note that double quotes are used all over the place in verify_users_pam.c .

olehermanse commented:
@xaxo05 Thank you for your work on this. Are you familiar with interactive rebase? (`git rebase -i`)

Could you squash the small fixups (reverts, style changes, whitespace) into their respective commits, so the commit history is more clear? From a quick look, these commits look like the ones which should remain:

```
Enable users promises on FreeBSD, use the utilities chpass, rmuser and ...

FreeBSD lacks fgetpwent() and fgetgrent(), implement them.
 
Create user using pw and use chpass to set user password on FreeBSD.
 
SetAccountLockExpiration using pw on FreeBSD.
 
Lock accounts on FreeBSD.
 
Use pw instead of usermod on FreeBSD.
 
Remove users on FreeBSD using pw.
```

Once you are done cleaning up using rebase, you'll have to force push here and we can do a final review and test.

If you are not comfortable with git, just tell me, and I can clean it up / squash it for you, to the best of my ability.

olehermanse commented:
@cf-bottom jenkins

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=4080)](https://ci.cfengine.com//job/pr-pipeline/4080/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4080/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4080/

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=4087)](https://ci.cfengine.com//job/pr-pipeline/4087/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4087/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4087/

olehermanse commented:
@cf-bottom test this in jenkins, with exotics, please.

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=4088)](https://ci.cfengine.com//job/pr-pipeline/4088/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4088/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4088/

vpodzime commented on cf-agent/verify_users_pam.c:
```diff
@@ -719,6 +719,19 @@ static bool ChangePassword(const char *puser, const char *password, PasswordForm
+#ifdef HAVE_PW
+    /* Accounts are locked by prepending "*LOCKED*" to the password
+     * hash on FreeBSD and other systems using pw. */
+    return (strstr(hash, "*LOCKED*") != NULL);
```
I think this is really OS-specific, not depending on an existence of some utility. Please use `#ifdef __FREEBSD__` here and for the similar checks in this commit.

olehermanse replied:
@vpodzime https://github.com/cfengine/core/pull/4002

olehermanse commented on cf-agent/verify_users_pam.c:
```diff
@@ -622,6 +630,45 @@ static bool ChangePasswordHashUsingLckpwdf(const char *puser, const char *passwo
+    Log(LOG_LEVEL_VERBOSE, "%s user '%s'. (command: '%s')", cap_action_msg, puser, cmd);
+
+    const int status = system(cmd);
+    if (!WIFEXITED(status) || WEXITSTATUS(status) != 0)
```
Apparently, on RHEL 5 this cannot be `const`, can you try to remove the `const` keyword here?

olehermanse replied:
Output:

```
11:34:32 verify_users_pam.c: In function 'ExecuteUserCommand':
11:34:32 verify_users_pam.c:647: error: assignment of read-only member '__in'
11:34:32 verify_users_pam.c:647: error: assignment of read-only member '__in'
11:34:32 gmake[2]: *** [verify_users_pam.lo] Error 1
11:34:32 gmake[2]: Target `all' not remade because of errors.
11:34:32 Making all in cf-execd
```

olehermanse replied:
@xaxo05 Did you click resolved on accident? We need the code to compile on RHEL 5 before merging it, so this needs to be fixed.

olehermanse commented:
@cf-bottom please build this in jenkins, with exotics.

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=4130)](https://ci.cfengine.com//job/pr-pipeline/4130/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4130/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4130/
